src/box/memtx_sort_data.c

```diff
@@ -0,0 +1,711 @@
+	writer->fp = NULL;
+	memset(writer->filename, 0, sizeof(writer->filename));
+	writer->entries = mh_memtx_sort_data_writer_entries_new();
+	writer->curr_entry = NULL;
+	writer->cardinality = 0;
+	writer->cardinality_offset = 0;
+	return writer;
+}
+
+void
+memtx_sort_data_writer_delete(struct memtx_sort_data_writer *writer)
+{
+	if (writer->fp != NULL)
+		fclose(writer->fp);
+	mh_memtx_sort_data_writer_entries_delete(writer->entries);
+	free(writer);
+}
+
+int
+memtx_sort_data_writer_create_file(struct memtx_sort_data_writer *writer,
+				   const char *snap_filename,
+				   struct vclock *vclock,
+				   const struct tt_uuid *instance_uuid,
+				   struct read_view *rv)
+{
+	/* Check a materialized file does not exist already. */
+	const char *filename = memtx_sort_data_filename(snap_filename);
+	if (access(filename, F_OK) == 0) {
+		errno = EEXIST;
+		diag_set(SystemError, "sort data file '%s'", filename);
+		return -1;
+	}
+
+	/* The in-progress file name. */
+	VERIFY((size_t)snprintf(writer->filename, sizeof(writer->filename),
+				"%s.inprogress", filename) <
+	       sizeof(writer->filename));
+
+	/* Open the file for write. */
+	say_info("saving memtx sort data `%s'", writer->filename);
+	writer->fp = fopen(writer->filename, "wb");
+	if (writer->fp == NULL) {
+		DIAG_SET(writer, "failed to open the file for write");
+		return -1;
+	}
+
+	/* Write the file header. */
+	if (print(writer,
+		  "SORTDATA\n"
+		  "1\n"
+		  "Version: %s\n"
+		  "Instance: %s\n"
+		  "VClock: %s\n\n",
+		  PACKAGE_VERSION,
+		  tt_uuid_str(instance_uuid),
+		  vclock_to_string(vclock)) != 0)
+		return -1;
+
+	/* The overall tuple count is to be updated later. */
+	if (get_offset(writer, &writer->cardinality_offset) != 0 ||
+	    print(writer, CARDINALITY_FMT, 0) != 0)
+		return -1;
+
+	/* Write the entry count to fill it later. */
+	long entries_offset;
+	if (get_offset(writer, &entries_offset) != 0 ||
+	    print(writer, ENTRIES_FMT, 0) != 0)
+		return -1;
+
+	/* Save the pointer to the first sort data header entry. */
+	if (get_offset(writer, &writer->next_header_entry_offset) != 0)
+		return -1;
+
+	/* Write dummy header entries to fill them later. */
+	uint32_t entry_count = 0;
+	struct space_read_view *space_rv;
+	read_view_foreach_space(space_rv, rv) {
+		if (space_rv->index_read_view_count <= 1)
+			continue;
+		for (uint32_t i = 0; i < space_rv->index_read_view_count; i++) {
+			print(writer, ENTRY_FMT, 0, 0, 0L, 0L, 0L);
+			entry_count++;
+		}
+	}
+
+	/* The final newline. */
+	if (print(writer, "\n") != 0)
+		return -1;
+
+	/* Write the calculated entry count and return. */
+	if (seek(writer, entries_offset, SEEK_SET) != 0)
+		return -1;
+	return print(writer, ENTRIES_FMT, entry_count);
+}
+
+int
+memtx_sort_data_writer_close_file(struct memtx_sort_data_writer *writer)
+{
+	if (seek(writer, writer->cardinality_offset, SEEK_SET) != 0 ||
+	    print(writer, CARDINALITY_FMT, (long long)writer->cardinality) != 0)
```

Comment:
I'm not sure it's safe to mix `seek` and `fprintf` and I don't understand how updating a line in the beginning of the file works without overwriting following data.

Reply:
My bad. I though that `seek` was a system call. Better rename it to avoid confusion.